### PR TITLE
Store attachments as intact files

### LINKED ISSUE #637: Attachment Should Have Filename Appended
Preserve the attached files in the original format.

### LINKED ISSUE #648: handle ambiguous leading slashes in s3 locations
S3 buckets are strictly speaking, k:v stores, where the key is a string.

AWS S3 and min.io (possibly others) allow folder-based navigation in their web UI's based on '/' separation of key components.

However, min.io does not seem to allow leading '/' components, and the folder view for AWS results in confusing '//' paths. e.g. top level '/folder' becomes '//folder', with '/' being a viewable entry at top level.

Therefore, I propose always trimming leading slashes from S3 objects to allow minio-S3 usage, of top level bucket directories (e.g. `location=''`) for blob storage, and prevent the confusing case for aws-S3.

One potential downside could be that existing S3 configurations using leading slashes may regress, or not be usable by other non-blob S3 access patterns (e.g. filepath objects). These sorts of regressions could be fixed by users bulk moving 'moving' the objects in '//folderlike' to 'folderlike' via web gui.

### LINKED ISSUE #650: Incorrect path reference
Error for external file opening gives error message referring to the wrong variable name: 

https://github.com/datajoint/datajoint-python/blob/74dfe4126b10b191c1dff47a179800b1434eeb64/datajoint/external.py#L140

### PR DESCRIPTION
**Pull request by dimitri-yatsenko**
* Fix #637 
* Fix #648
* Fix #650

### CODE REVIEW

**Comment by guzman-raphael** [comment]
Can you please remove python 3.4 from `.travis.yml` if we no longer plan on testing against it?

**Comment by guzman-raphael** [comment]
Please merge in the latest from `dev` to include new testing structure.

File: tests/test_filepath.py [hunk 1]
```diff
@@ -18,34 +19,34 @@ def test_filepath(store="repo"):
 
     # create a mock file
     relpath = 'one/two/three'
-    os.makedirs(os.path.join(stage_path, relpath), exist_ok=True)
-    managed_file = os.path.join(stage_path, relpath, filename)
+    managed_file = Path(stage_path, relpath, filename)
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: This test failed on WIN10. For example, the result of `managed_file` was `C:\\Users\\rguzman\\AppData\\Local\\Temp\\tmpb3gvp0ht\\one\two\threepicture.dat`.

File: tests/test_filepath.py [hunk 2]
```diff
@@ -18,34 +19,34 @@ def test_filepath(store="repo"):
...
+    with managed_file.open('rb') as f:
         synced_data = f.read()
     assert_equal(data, synced_data)
 
     # cleanup
-    ext.delete()
+    ext.delete(delete_external_files=True)
 
 
 def test_filepath_s3():
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: This test failed on WIN10. For example, the result of `managed_file` was `dj/repo/one\two\three\picture.dat`.

File: tests/test_filepath.py [hunk 3]
```diff
@@ -56,14 +57,13 @@ def test_filepath_s3():
 def test_duplicate_upload(store="repo"):
     ext = schema.external[store]
     stage_path = dj.config['stores'][store]['stage']
-    filename = 'plot.dat'
     relpath = 'one/two/three'
-    os.makedirs(os.path.join(stage_path, relpath), exist_ok=True)
-    managed_file = os.path.join(stage_path, relpath, filename)
-    with open(managed_file, 'wb') as f:
+    managed_file = Path(stage_path, relpath, 'plot.dat')
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: This test failed on WIN10. For example, the result of `managed_file` was `dj/repo/one\two\three\plot.dat`.

File: tests/test_filepath.py [hunk 4]
```diff
@@ -75,16 +75,15 @@ def test_duplicate_error(store="repo"):
     """ syncing duplicate non-matching file should fail """
     ext = schema.external[store]
     stage_path = dj.config['stores'][store]['stage']
-    filename = 'thesis.dat'
     relpath = 'one/two/three'
-    os.makedirs(os.path.join(stage_path, relpath), exist_ok=True)
-    managed_file = os.path.join(stage_path, relpath, filename)
-    with open(managed_file, 'wb') as f:
+    managed_file = Path(stage_path, relpath, 'thesis.dat')
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: This test failed on WIN10. For example, the result of `managed_file` was `dj/repo/one\two\three\plot.dat`.

File: tests/test_filepath.py [hunk 5]
```diff
@@ -56,14 +57,13 @@ def test_filepath_s3():
...
+    managed_file.parent.mkdir(parents=True, exist_ok=True)
+    with managed_file.open('wb') as f:
         f.write(os.urandom(300))
-    ext.fput(managed_file)
-    ext.fput(managed_file)   # this is fine because the file is the same
+    ext.upload_filepath(managed_file)
+    ext.upload_filepath(managed_file)  # this is fine because the file is the same
 
 
 def test_duplicate_upload_s3():
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: This test failed on WIN10. For example, the result of `managed_file` was `dj/repo/one\two\three\plot.dat`.

File: tests/test_filepath.py [hunk 6]
```diff
@@ -75,16 +75,15 @@ def test_duplicate_error(store="repo"):
...
-    ext.fput(managed_file)
-    with open(managed_file, 'wb') as f:
+    ext.upload_filepath(managed_file)
+    with managed_file.open('wb') as f:
         f.write(os.urandom(300))
-    ext.fput(managed_file)  # this should raise exception because the file has changed
+    ext.upload_filepath(managed_file)  # this should raise exception because the file has changed
 
 
 def test_duplicate_error_s3():
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: This test failed on WIN10. For example, the result of `managed_file` was `dj/repo/one\two\three\thesis.dat`.

File: tests/test_filepath.py [hunk 7]
```diff
@@ -93,32 +92,30 @@ def test_duplicate_error_s3():
 
 def test_filepath_class(table=Filepath(), store="repo"):
     stage_path = dj.config['stores'][store]['stage']
-    filename = 'attachment.dat'
-
     # create a mock file
     relative_path = 'one/two/three'
-    os.makedirs(os.path.join(stage_path, relative_path), exist_ok=True)
-    managed_file = os.path.join(stage_path, relative_path, filename)
+    managed_file = Path(stage_path, relative_path, 'attachment.dat')
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: This test failed on WIN10. For example, the result of `managed_file` was `C:\\Users\\rguzman\\AppData\\Local\\Temp\\tmpb3gvp0ht\\one\two\threeattachment.dat`.

File: tests/test_filepath.py [hunk 8]
```diff
@@ -127,7 +124,7 @@ def test_filepath_class(table=Filepath(), store="repo"):
     assert_true(table.external[store])
 
     # delete from external table
-    table.external[store].delete()
+    table.external[store].delete(delete_external_files=True)
 
 
 def test_filepath_class_again():
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: - This test failed on WIN10 at `test_filepath_class_again`. Error details: `datajoint.errors.DataJointError: A different version of 'one\two\three\attachment.dat' has already been placed`
- This test failed on WIN10 at `test_filepath_class_s3`. For example, the result of `managed_file` was `dj/repo/one\two\three\attachment.dat`.
- This test failed on WIN10 at `test_filepath_class_s3_again`. For example, the result of `managed_file` was `dj/repo/one\two\three\attachment.dat`.

File: tests/test_filepath.py [hunk 9]
```diff
@@ -147,20 +144,18 @@ def test_filepath_class_s3_again():
...
-        os.makedirs(os.path.join(stage_path, relative_path), exist_ok=True)
-        managed_file = os.path.join(stage_path, relative_path, filename)
-        with open(managed_file, 'wb') as f:
-            f.write(contents)   # same contents in all the files
+        relative_path = Path(*random.sample(('one', 'two', 'three', 'four'), k=3))
+        managed_file = Path(stage_path, relative_path, 'file.dat')
+        managed_file.parent.mkdir(parents=True, exist_ok=True)
+        with managed_file.open('wb') as f:
+            f.write(contents)  # same in all files
         table.insert1((i, managed_file))
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: This test failed on WIN10. Error details: `datajoint.errors.DuplicateError: ("Duplicate entry '1' for key 'PRIMARY'", 'To ignore duplicate entries in insert, set skip_duplicates=True')`

File: tests/test_filepath.py [hunk 10]
```diff
@@ -169,16 +164,9 @@ def test_filepath_cleanup(table=Filepath(), store="repo"):
...
-    unused_files = list(ext.get_untracked_filepaths())
-    assert_true(0 < len(unused_files) <= m)
-
-    # check no more untracked files
-    ext.clean_filepaths()
-    assert_false(bool(list(ext.get_untracked_filepaths())))
-
 
 def test_filepath_cleanup_s3():
     """test deletion of filepath entries from external table """
```
Comment (guzman-raphael) [inline tests/test_filepath.py]: This test failed on WIN10. For example, the result of `managed_file` was ` dj/repo/four\three\one\file.dat`.

File: .travis.yml [hunk 11]
```diff
@@ -69,4 +69,5 @@ jobs:
       language: shell
       script:
         - sudo pip install python-coveralls
-        - coveralls
\ No newline at end of file
+        - coveralls
+>>>>>>> bff58baac8da58b43be95ba1c58174cb7721411f
```
Comment (guzman-raphael) [inline .travis.yml]: This line should be removed...

File: tests/test_filepath.py (hunk 9 above)
Comment (dimitri-yatsenko) [reply tests/test_filepath.py]: this happens sometimes when the previous test run was terminated before cleanup

**Pull request merged**